SO-4227: migrate 6.x donation logic

cmark opened the pull request.

cmark committed:
SO-4227: add filterConflicts method to RevisionConflictProcessor...

...to support filtering of conflicts by domain specific logic before
reporting.

cmark committed:
SO-4227: fix issue with ApiException.withAdditionalInfo(Map) method

cmark committed:
SO-4227: introduce Json.object() and Json.array()

Migrate test cases and test fixtures to use Json.object instead of hard
to read (and maintain) Map instances.

cmark committed:
SO-4227: respond with HTTP 409 when upgrade detects conflicts

`conflicts` property should list all detected conflicts.

cmark committed:
SO-4227: support detection of donated SNOMED CT components

Migrate first five extension upgrade test cases from 6.x.

cmark committed:
SO-4227: make sure we always use today or later dates for new versions

cmark committed:
SO-4227: add more test cases and functionality to upgrade donation...

...processing

cmark committed:
SO-4227: delay registering externally marked revised revisions during...

...upgrade

apeteri commented on commons/com.b2international.index/src/com/b2international/index/revision/StagingArea.java:
```diff
@@ -786,6 +812,10 @@ void merge(RevisionBranchRef fromRef, RevisionBranchRef toRef, boolean squash, R
 			
 			newRevisions.forEach(rev -> {
+				// skip new objects that are already marked as revised on merge source, content that is present on target should take place instead
+				if (externalRevisionsToReviseOnMergeSource.containsEntry(rev.getClass(), rev.getId())) {
```
Should there be a similar check for `updatedRevisions` below?

cmark replied:
We will see, we can add it earlier, but for now, there is no test case for it.

cmark committed:
SO-4227 #resolve

Migrate all remaining donation 6.x test cases.
Remove obsolete SnomedExtensionDowngradeTest and
SnomedExtensionVersioningTest test cases.

cmark committed:
SO-4227 improve performance by not loading donated content at all...

...when upgrading

cmark committed:
SO-4227: slightly improve performance by freeing up some memory...

...during upgrade
